### PR TITLE
enhancement(ci): Add Github Workflow to publish custom builds from a branch.

### PR DESCRIPTION
**Pull request by neuronull**
This introduces a new GH Workflow for publishing custom builds to s3 and docker, that borrows much from the Nightly workflow.
The idea is to be able to use it to create builds for users/customers on the fly. This can be useful for when providing fix candidates that haven't been merged in yet, which can be the case when we are not able to locally reproduce an issue but think we know the root cause.

The workflow leverages the environment variable `CHANNEL` which is picked up on in the `vdev release channel` command, which sets the version string to a something like `<base_version>-custom-<git shorthash>` if the channel is "custom" . For example, `0.29.0.custom.9da843f`. This is then used in the build artifact labels, and docker/s3 organization.

See 

https://packages.timber.io/vector/custom/0.29.0.custom.9da843f/

https://hub.docker.com/repository/docker/timberio/test-custom-builds/tags?page=1&ordering=last_updated

**NOTE**:
- the docker repo was temporarily changed so as to not create noise in the main repo. It will be reverted back to the normal repo before merging.
- we skip homebrew, cloudsmith and github exporting for custom builds.

### CODE REVIEW

File: .github/workflows/nightly.yml
Comment (neuronull) [inline .github/workflows/nightly.yml:13]: Wondering if anywhere else downstream is expecting this to be capitalized ...

File: vdev/src/commands/release/channel.rs
Comment (neuronull) [inline vdev/src/commands/release/channel.rs:5]: update this comment to be more accurate

File: vdev/src/util.rs
Comment (neuronull) [inline vdev/src/util.rs:40]: update

File: .github/workflows/nightly.yml
Comment (neuronull) [reply .github/workflows/nightly.yml:13]: Not finding any.

File: scripts/build-docker.sh
Comment (spencergilbert) [inline scripts/build-docker.sh:19]: Commenting to ensure this isn't forgotten

**Review by spencergilbert — APPROVED** [review]
This seems correct based on my github actions experience, but I'm not an expert there 😅

File: scripts/build-docker.sh
Comment (bruceg) [inline scripts/build-docker.sh:19]: Still TODO?

File: vdev/src/commands/version.rs
Comment (bruceg) [inline vdev/src/commands/version.rs:16]: These changes should conflict with what @jonathanpv just merged here.

File: .github/actions/spelling/expect.txt
Comment (bruceg) [inline .github/actions/spelling/expect.txt:350]: What needs this added?

**Review by bruceg — CHANGES_REQUESTED** [review]
The vdev bits make sense, but are going to conflict with what I believe was just merged.

File: .github/actions/spelling/expect.txt
Comment (neuronull) [reply .github/actions/spelling/expect.txt:350]: A code comment in a shell script for build msi package:

```
# "0.29.0.custom.a28ecdc" -> "0.29.0"
```

, it flagged "ecdc"

I liked having a specific example of a shorthash as opposed to "<version>.custom.<git shorthash>" and was resisting changing my comment because of a false flag of a spell checker.

File: vdev/src/commands/version.rs
Comment (neuronull) [reply vdev/src/commands/version.rs:16]: Shucks... ok I'll merge master in. Thanks for the heads up.

File: scripts/build-docker.sh
Comment (neuronull) [reply scripts/build-docker.sh:19]: It'll be the last thing I revert back as mentioned in the opening comment.

File: vdev/src/commands/version.rs
Comment (neuronull) [reply vdev/src/commands/version.rs]: addressed in https://github.com/vectordotdev/vector/pull/16281/commits/7e3361b4b6e25e163c002fd76548ca7bfa4e8def

File: .github/actions/spelling/expect.txt
Comment (neuronull) [reply .github/actions/spelling/expect.txt:350]: I can double check. The only suggestions I saw in the error output it gave, were to ignore whole files, or add exceptions for words.

File: vdev/src/commands/release/channel.rs
Comment (neuronull) [reply vdev/src/commands/release/channel.rs:7]: AFAICT, the `cargo vdev release channel` command is only used within the build automation we have. If that is the case, I think we can just use `MODE` like you suggest. My concern was if that is used elsewhere, where we wouldn't have MODE explicitly set.

File: .github/actions/spelling/expect.txt
Comment (spencergilbert) [reply .github/actions/spelling/expect.txt:350]: I believe you could add a pattern to `patterns.txt`?

Comment (neuronull) [reply .github/actions/spelling/expect.txt]: attempted to address in https://github.com/vectordotdev/vector/pull/16281/commits/e354b138153f40e17347079ec12ad321acad4602

File: vdev/src/commands/release/channel.rs
Comment (neuronull) [reply vdev/src/commands/release/channel.rs]: Attempted to address in https://github.com/vectordotdev/vector/pull/16281/commits/048c92feb42054b02f268458875b47c410c97de2

**Comment by jszwedko** [comment]
> > It's not a huge deal, but is there a reason you include the date in the docker tags but not in the AWS S3 artifacts? I would have expected them to be consistent.
> 
> Good observation. I think in earlier weeks I had a mental flag to look at that but forgot about it. There wasn't a specific reason.
> 
> We actually have a few options... In s3, I was considering organizing the builds by a $DATE folder, within that being the folder for that .custom.. But that seemed like it would be harder to find things. Do you agree?
>
> I could just prefix the $DATE into both the s3 folder for the custom build, or just the artifacts within the folder. So I guess there are three options :) I don't have a huge preference either way. Maybe adding it to both the folder, if I had to choose.

Given that I think users (and us) will typically be looking for custom builds by commit rather than date I'd suggest just dropping the dates from the docker tags so they match up with your package builds.

File: vdev/src/commands/release/channel.rs
Comment (jszwedko) [reply vdev/src/commands/release/channel.rs]: Looks good, thanks!

File: .github/workflows/publish.yml
Comment (jszwedko) [inline .github/workflows/publish.yml:43]: Maybe we should just move this comment up to the `mode` argument at the head of the workflow? Nit but you could also move this env var into an `env: ` block as you see elsewhere in this file.

File: build.rs
Comment (jszwedko) [inline build.rs:101]: When would this fail?

File: scripts/ci-generate-publish-metadata.sh
Comment (jszwedko) [inline scripts/ci-generate-publish-metadata.sh:24]: Why is this needed now?

File: scripts/package-msi.sh
Comment (jszwedko) [inline scripts/package-msi.sh:26]: Why do we need to strip this?

File: vdev/src/app.rs
Comment (jszwedko) [inline vdev/src/app.rs:65]: Ah I see. Do you have a link to one of these failures? I'm a bit confused as to what would cause it.

**Review by jszwedko — COMMENTED** [review]
Awesome! Just a few more comments/questions.

File: vdev/src/app.rs
Comment (neuronull) [reply vdev/src/app.rs:65]: Mhm. Basically it's a git permissions thing that happens within containers. We have a command run in the publish workflow to handle it for some of those cases, but in this failure it's because it is failing from the call out to the Makefile targets that do the packaging, which invokes the docker run, hence my changes to pass forward that env var in the Makefile.

https://github.com/vectordotdev/vector/actions/runs/4428339593/jobs/7767355987

File: scripts/ci-generate-publish-metadata.sh
Comment (neuronull) [reply scripts/ci-generate-publish-metadata.sh:24]: Same answer as https://github.com/vectordotdev/vector/pull/16281#discussion_r1137833090

File: scripts/package-msi.sh
Comment (neuronull) [reply scripts/package-msi.sh:26]: This change was because of this failure, which I sort of reference on line 22 (chose that line because CHANNEL is only needed because of this logic).

```
vector.wxs
C:\a\vector\vector\target\msi-x64\vector.wxs(6) : error CNDL0108 : The Product/@Version attribute's value, '0.29.0.custom.1bce710', is not a valid version.  Legal version values should look like 'x.x.x.x' where x is an integer from 0 to 65534.
```

https://github.com/vectordotdev/vector/actions/runs/4367571927/jobs/7639109583

File: build.rs
Comment (neuronull) [reply build.rs:101]: The `output()` call actually executes the command in a subprocess so there could be a lot of reasons it could fail I guess. Since this is in a critical code path for compilation I figured this would be safer than doing an `unwrap()` .

But I'm not _expecting_ it to fail. I could add a clarifying comment if that would be helpful.

File: scripts/ci-generate-publish-metadata.sh
Comment (neuronull) [reply scripts/ci-generate-publish-metadata.sh:24]: We could compute it later on closer to where it is needed, but this seemed pretty harmless and saves the repeated effort.

File: vdev/src/app.rs
Comment (neuronull) [reply vdev/src/app.rs:65]: My comment in the `build.rs` file actually does a little better job explaining the same reasoning that this one is trying to do. I can transpose that one here.

File: scripts/package-msi.sh
Comment (jonathanpv) [reply scripts/package-msi.sh:21]: Implemented in https://github.com/vectordotdev/vector/pull/16748/files#diff-3aa0b9b03ec0a4a4861c47ab7de668a62095b8669d0fd592df5cb84c7521a05eR31-R39

File: vdev/src/app.rs
Comment (jszwedko) [reply vdev/src/app.rs:65]: We do actually have a precedent of setting `safe.directory` to workaround that issue: https://github.com/vectordotdev/vector/blob/master/scripts/environment/prepare.sh#L4

Maybe we could do that here too? Or call that same prepare.sh script?

File: scripts/package-msi.sh
Comment (jszwedko) [reply scripts/package-msi.sh:26]: Ah, gotcha. Maybe we could leave that as a comment?

File: vdev/src/app.rs
Comment (neuronull) [reply vdev/src/app.rs:65]: Good idea. I was working around the issue by adding the sha to an env var. But this seems like a better approach. My first attempt was to add the safe directory in the workflow job, but that didn't work since the issue was in the container.
But we can still run that command while we are in the scripts executing within the container.

Comment (neuronull) [reply vdev/src/app.rs:65]: @jszwedko - one thing I just thought of as I'm implementing this. Regardless of whether we compute the SHA and store in a env var, or mark the repo as safe, it strikes me that this could introduce a problem for anyone who is building vector from source, within a container. I guess the only way around that would be to add logic into `build.rs` , which would call the git command to add safe directory, for the current working directory.

Do you think this is a valid concern,?

Comment (neuronull) [reply vdev/src/app.rs:65]: The other comment raised about explicitly throwing error if the git command fails, is essentially what would otherwise happen, if we didn't do anything.

Comment (neuronull) [reply vdev/src/app.rs:65]: We discussed offline and decided that this is a marginal concern and that it is reasonable to expect the permissions to be set, but I will comment it.

**Review by spencergilbert — COMMENTED** [review]
Nothing jumped out to me, other than needing to revert the REPO

File: vdev/src/commands/mod.rs
Comment (bruceg) [inline vdev/src/commands/mod.rs:133]: This macro appears to be identical to the above except for this line. Any chance of merging the two, maybe by using a null expression where this would go?

File: vdev/src/util.rs
Comment (bruceg) [inline vdev/src/util.rs:44]:
```suggestion
    if matches!(std::env::var("CI"), Ok("true")) {
```

File: vdev/src/util.rs
Comment (bruceg) [inline vdev/src/util.rs:61]: I'd prefer this to be named something other than just "mode" and `$MODE`. What kind of mode? We refer to it as "release channel" elsewhere, maybe use that?

File: build.rs
Comment (spencergilbert) [reply build.rs:195]: I thought this was an enforced clippy lint - or is it not until a newer rust version?

File: build.rs
Comment (jszwedko) [inline build.rs:197]:
```suggestion
        .expect("git hash detection failed");
```

File: Makefile
Comment (jszwedko) [inline Makefile:530]: Do we still need to pass in `VERSION` here now that the CLI is doing the detection again?

File: vdev/src/util.rs
Comment (neuronull) [reply vdev/src/util.rs:45]: My thought process was that the CI build workflows are tightly coupled to vdev already (the Makefile targets that the CI workflow invokes directly call `vdev package <...>`), and that this was a more centralized location for the logic, than a new step in each of the specific packaging jobs, because we also have to do it before running the vdev "version" subcommand.
The command is gated on CI env var so it won't be executed in local vdev usage.

File: vdev/src/commands/mod.rs
Comment (neuronull) [reply vdev/src/commands/mod.rs]: attempted to resolve in [2b0da0b](https://github.com/vectordotdev/vector/pull/16281/commits/2b0da0b0b2ce4cfdbd67b51e77063b67b0023959)

File: Makefile
Comment (neuronull) [reply Makefile]: seems to have worked :)

File: build.rs
Comment (neuronull) [reply build.rs:193]: yeah this is being enforced by clippy

**Review by bruceg — APPROVED** [review]
I'm on the fence about if the extra parameter to the `script_wrapper` macro is a big deal or not, but the rest is fine, so I'm approving unless you really want to simplify that one.

File: vdev/src/commands/mod.rs
Comment (neuronull) [reply vdev/src/commands/mod.rs:97]: 😮 I just fell in love with macros

File: .github/workflows/custom_builds.yml
Comment (neuronull) [reply .github/workflows/custom_builds.yml:8]: Not sure, it was a copy paste. I will try removing it and see.

Comment (neuronull) [reply .github/workflows/custom_builds.yml:8]: apparently that is a requirement 😆

File: vdev/src/util.rs
Comment (neuronull) [reply vdev/src/util.rs]: chatted with Jesse about this out of band. I will try pulling this out of vdev into a pre step in CI, and if issues arise which led to the placement in vdev I can surface that.

We may decide to make vdev more CI aware but that doesn't have to be in scope for this endeavor.

File: .github/workflows/custom_builds.yml
Comment (jszwedko) [inline .github/workflows/custom_builds.yml:4]: Just a reminder to swap this out for `workflow_dispatch` before merging.

File: .github/workflows/custom_builds.yml
Comment (jszwedko) [inline .github/workflows/custom_builds.yml:7]:
```suggestion
  Custom:
```

File: build.rs
Comment (jszwedko) [inline build.rs:195]: I think this `expect` will actually swallow the underlying error. Maybe the `map_err` should `eprintln!()`?

File: vdev/src/commands/build/publish_metadata.rs
Comment (jszwedko) [inline vdev/src/commands/build/publish_metadata.rs:22]: It stuck out to me that this only reads `VERSION` while `app.rs` also reads `VECTOR_VERSION`. Is there a reason they are different? Should there maybe be a helper function they both use that reads both along with the fallback to reading it from `git`?

File: Makefile
Comment (jszwedko) [inline Makefile:530]: I'm not actually seeing where `VECTOR_VERSION` is being set. Am I missing something? Or is this just for debugging purposes?

**Review by jszwedko — COMMENTED** [review]
Thanks for the updates! Just a few last questions (hopefully 😅 ).

File: Makefile
Comment (neuronull) [reply Makefile:530]: it is exported by the `publish.yml` jobs, which take it from the result of the `vdev generate publish metadata`

File: vdev/src/commands/build/publish_metadata.rs
Comment (neuronull) [reply vdev/src/commands/build/publish_metadata.rs:22]: Good spot. The publish metadata command is run (AFAICT) only at the beginning of the `publish.yml` workflow, and this workflow is what provides `VECTOR_VERSION` , so I don't think it would improve anything functionally, by adding it here too (or in a central helper function).
But I guess it wouldn't hurt to do that either (at least in this current state of the workflows), if it would dispel confusion.
I guess an alternative is adding a code comment.

What do you think makes the most sense?

Comment (jszwedko) [reply vdev/src/commands/build/publish_metadata.rs:22]: Hmm, yeah, I could just see it causing confusion. At the least, a code comment would help.

Comment (neuronull) [reply vdev/src/commands/build/publish_metadata.rs]: 👍 I extracted that into a helper.

**Comment by neuronull** [comment]
Everything looks good with the latest changes. I will change back the trigger type.
Not sure why the Netfily deploys failed 🤔 
But assuming CI is green after this change, this should be ready to merge!

**Review by jszwedko — APPROVED** [review]

**Pull request merged**
